include/rdma/fi_collective.h

```diff
@@ -72,6 +72,12 @@ struct fi_collective_attr {
 	uint64_t 		mode;
 };
 
+struct fi_collective_join_attr {
+	uint64_t		bitmap_ops_wanted;	/* 1 << FI_BARRIER, etc */
+	uint64_t		bitmap_ops_supported;	/* 1 << FI_BARRIER, etc */
+	void			*s_context;		/* session context */
```

Comment:
I don't understand the need for s_context.  Why can't the app manage this via the context param?

There's already a query call to return what ops are supported.  The app should have checked that prior to requesting these ops.  The purpose for bitmap_ops_supported isn't clear.

The attributes for a collective join could be significantly more complex than this.  There are a variety of ops, each of which has a number of datatypes, different number of members, etc.  If a more generic join call is being defined, it should be extensible to capture more details without being unusable.

Note that if the only need is to identify which ops the join should support that's actually captured today through the av_set creation.

Reply:
Sure the s_context can be passed vi context by defining a structure like how the peer API does. However, since we already have the attr structure here, add s_context to attr is simpler.

I missed the av_set flags part. Yes the main purpose of the new join is to pass in the ops needed. May be able to drop this commit.